crates/runtime/README.md

```diff
@@ -0,0 +1,662 @@
+of capabilities to the guest Wasm code, allowing it to interact with the outside
+world in a controlled and secure manner. The primary purpose of host functions
+is to extend the functionality of the Wasm module beyond its inherent
+limitations while maintaining strict control over its access to system resources
+and external services.
+
+At the core of the host function implementation is the `VMHostFunctions` struct,
+which encapsulates the runtime's logic and provides a safe interface for the
+guest code to interact with. This struct is created using self-references,
+allowing it to maintain a consistent view of the execution environment
+throughout the lifetime of a function call.
+
+One of the fundamental responsibilities of host functions is memory management.
+They provide methods for reading from and writing to the guest's memory space,
+ensuring that all memory accesses are bounds-checked and secure. This includes
+functions for reading raw bytes from memory as well as higher-level operations
+like retrieving UTF-8 strings. By mediating all memory interactions, the host
+functions prevent potential security vulnerabilities that could arise from
+unchecked memory access.
+
+Register operations form another critical aspect of the host functions. They
+allow the guest code to store and retrieve data in a temporary storage system,
+which is particularly useful for passing information between different parts of
+the Wasm module or for storing intermediate results. The host functions ensure
+that all register operations adhere to the size and count limits specified in
+`VMLimits`, preventing resource abuse and maintaining the integrity of the
+execution environment.
+
+Storage interactions are also facilitated through host functions. These
+functions provide a key-value storage interface, allowing the guest code to read
+from and write to a persistent storage system. By channeling all storage
+operations through these functions, the runtime can enforce access controls,
+manage resource usage, and potentially implement advanced features like rollback
+or versioning.
+
+Logging and event emission are crucial for debugging and monitoring Wasm
+execution. The host functions provide methods for the guest code to log messages
+and emit events, which are collected by the runtime for later analysis. These
+functions ensure that logging and event emission adhere to the specified limits,
+preventing potential denial-of-service attacks through excessive logging.
+
+One of the more powerful capabilities provided by the host functions is the
+ability to make external HTTP requests. The `fetch()` function allows the guest
+code to interact with external services, opening up possibilities for complex,
+interconnected applications. However, this function is implemented with careful
+consideration for security, potentially including restrictions on allowed URLs
+or rate limiting to prevent abuse.
```

Comment:
we're unable to do any of these generically without being opinionated, but higher-level libraries like `calimero-sdk-near` unsafely use the `calimero_sdk::env::ext::fetch` method and can apply their service-specific behaviour

so this is out of scope for the runtime

https://github.com/calimero-network/core/blob/a1376c7f823397015fa4d45b7542d6a05c95a1a8/crates/sdk/src/env/ext.rs#L5

Reply:
Okay, that's fair - I did say "potentially", but part of the problem is I don't quite know yet where the line is drawn between some parts 😄 So I will happily remove this 👍